File: java8/src/main/java/com/shekhargulati/ninetynine_problems/_01_lists/P10.java
package com.shekhargulati.ninetynine_problems._01_lists;

import java.util.AbstractMap.SimpleEntry;
import java.util.List;

import static java.util.stream.Collectors.toList;

/**
 * <b> (*) Run-length encoding of a list.</b>
 */
public class P10 {

    public static <T> List<SimpleEntry<Integer, T>> encode(List<T> list) {
        return P09.pack(list).stream().map(l -> new SimpleEntry<>(l.size(), l.get(0))).collect(toList());
    }
}

File: scala/src/main/scala/com/shekhargulati/ninetynine_problems/_01_lists/P10.scala
package com.shekhargulati.ninetynine_problems._01_lists

/**
  * <b> (*) Run-length encoding of a list.</b>
  * {{{
  *   encode(List("a", "a", "a", "a", "b", "c", "c", "a", "a", "d", "e", "e", "e", "e"))
  * }}}
  */
object P10 {

  def encode[T](list: List[T]): List[(Int, T)] = {
    P09.pack_span(list).map(l => (l.size, l.head))
  }

}
